kubelet: Move Lock Contention flags to configuration

ipochi opened the pull request:
This commit moves the `--exit-on-lock-contention` and `--lock-file` kubelet flags to Kubelet Configuration.

This PR is built on the following PRs:

Corresponding E2E test PRs : https://github.com/kubernetes/kubernetes/pull/103608, https://github.com/kubernetes/kubernetes/pull/104334, https://github.com/kubernetes/kubernetes/pull/108563

Corresponding Job in test-infra :
https://github.com/kubernetes/test-infra/blob/e684255cc8701ef97b6832e3daadb6841c00cc65/config/jobs/kubernetes/sig-node/containerd.yaml#L1315-#L1343

Signed-off-by: Imran Pochi <imran@kinvolk.io>

#### What type of PR is this?

/kind kubelet
/kind cleanup
/kind sig-node

#### What this PR does / why we need it:

Moves the `--exit-on-lock-contention` and `--lock-file` kubelet flags to Kubelet Configuration.

#### Which issue(s) this PR fixes:

References #86843

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

The Kubelet's `--lock-file` and `--exit-on-lock-contention` flags are now available via the Kubelet config file field `lockFile` and `exitOnLockContention`


/cc: @knabben @rata

Commit:
kubelet: Move Lock Contention flags to configuration

This commit moves the `--exit-on-lock-contention` and `--lock-file`
kubelet flags to Kubelet Configuration.

This PR is built on the following PRs:

Corresponding E2E test PRs : #103608, #104334, #108563

Corresponding Job in test-infra :
https://github.com/kubernetes/test-infra/blob/e684255cc8701ef97b6832e3daadb6841c00cc65/config/jobs/kubernetes/sig-node/containerd.yaml#L1315-#L1343

Signed-off-by: Imran Pochi <imran@kinvolk.io>

ipochi commented:
/test pull-kubernetes-integration

mtaufen commented on pkg/kubelet/apis/config/fuzzer/fuzzer.go:
```diff
@@ -110,6 +110,8 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 			}
 			obj.EnableSystemLogHandler = true
 			obj.MemoryThrottlingFactor = utilpointer.Float64Ptr(rand.Float64())
+			obj.ExitOnLockContention = true
```
I think we only needed to provide fixed values for fields where the defaulter will change it from the zero-value to some other value (which messes up the fuzzed round trip). Since the defaults for these fields are the zero-value, I don't think we have to set them here.

ipochi replied:
@mtaufen Hey sorry, this work got delayed due to the pre-requisite e2e test.

I'm not completely sure about this but reading the comment on this [line](https://github.com/ipochi/kubernetes/blob/b5be8e42467966bdf8f2c33674f430e0d857cc28/pkg/kubelet/apis/config/fuzzer/fuzzer.go#L39) makes your point valid.

mtaufen commented:
/triage accepted
/priority important-longterm

pacoxu commented:
/release-note-edit
```release-note
The Kubelet's `--lock-file` and `--exit-on-lock-contention` flags are now available via the Kubelet config file field `lockFile` and `exitOnLockContention`
```

knabben commented:
@ipochi closing https://github.com/kubernetes/kubernetes/pull/91877 FYI

rata commented:
It will be great to move this forward. Besides the rebase, what was missing here?

/remove-lifecycle stale

ipochi commented:
> It will be great to move this forward. Besides the rebase, what was missing here?
> 
> /remove-lifecycle stale

@rata Need to get this PR merged. I'll be attending the SIG call  Every Wednesday 10AM PT to drive this forward.

ipochi commented:
/retest-required

ipochi commented:
@SergeyKanzhelev @endocrimes @ehashman 

Rebased the PR and addressed the feedback from @mtaufen.

One of the pre-requisites was to have an E2E test for this flags, which is now merged in the master, meaning the final move of the flags to KubeletConfiguration can proceed ahead.

Please let me know the next steps in order to get this merged.

/cc @rata @iaguis

rata commented:
@ipochi just for convenience, can you link the PR that is now merged?

ipochi commented:
> @ipochi Thanks! Just for convenience, can you link the PR that is now merged?

Ah, its in the commit message, forgot that the PR description needs to be updated as well.
Now the reference to the E2E test PRs and the job config is in the updated PR description.

rata commented:
@ipochi tests are failing and they seem on spot. See the logs for more details, but here there is a quick summary:

```
Mar 15 10:34:30.487: INFO: observed event type MODIFIED
Mar 15 10:34:30.489: INFO: observed event type MODIFIED
Mar 15 10:34:30.489: INFO: observed event type MODIFIED
Mar 15 10:34:30.489: INFO: observed event type MODIFIED
Mar 15 10:34:30.489: INFO: observed event type MODIFIED
Mar 15 10:35:30.486: FAIL: failed to see DELETED event
Unexpected error:
    <*errors.errorString | 0xc000204240>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred
Full Stack Trace
k8s.io/kubernetes/test/e2e/common/node.glob..func14.13()
	test/e2e/common/node/pods.go:1054 +0x166e
k8s.io/kubernetes/test/e2e.RunE2ETests(0x245d257)
	test/e2e/e2e.go:130 +0x687
k8s.io/kubernetes/test/e2e.TestE2E(0x23d4919) 
....
 • Failure [89.249 seconds]
[sig-node] Pods
test/e2e/common/node/framework.go:23
  should run through the lifecycle of Pods and PodStatus [Conformance] [It]
  test/e2e/framework/framework.go:633
  Mar 15 10:35:30.486: failed to see DELETED event
              
  Unexpected error:
      <*errors.errorString | 0xc000204240>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  occurred
  test/e2e/common/node/pods.go:1054
------------------------------
{"msg":"FAILED [sig-node] Pods should run through the lifecycle of Pods and PodStatus [Conformance]","total":-1,"completed":1,"skipped":25,"failed":1,"failures":["[sig-node] Pods should run through the lifecycle of Pods and PodStatus [Conformance]"]} 
```

ipochi commented:
> @ipochi tests are failing and they seem on spot. See the logs for more details, but here there is a quick summary:
> 
> ```
> Mar 15 10:34:30.487: INFO: observed event type MODIFIED
> Mar 15 10:34:30.489: INFO: observed event type MODIFIED
> Mar 15 10:34:30.489: INFO: observed event type MODIFIED
> Mar 15 10:34:30.489: INFO: observed event type MODIFIED
> Mar 15 10:34:30.489: INFO: observed event type MODIFIED
> Mar 15 10:35:30.486: FAIL: failed to see DELETED event
> Unexpected error:
>     <*errors.errorString | 0xc000204240>: {
>         s: "timed out waiting for the condition",
>     }
>     timed out waiting for the condition
> occurred
> Full Stack Trace
> k8s.io/kubernetes/test/e2e/common/node.glob..func14.13()
> 	test/e2e/common/node/pods.go:1054 +0x166e
> k8s.io/kubernetes/test/e2e.RunE2ETests(0x245d257)
> 	test/e2e/e2e.go:130 +0x687
> k8s.io/kubernetes/test/e2e.TestE2E(0x23d4919) 
> ....
>  • Failure [89.249 seconds]
> [sig-node] Pods
> test/e2e/common/node/framework.go:23
>   should run through the lifecycle of Pods and PodStatus [Conformance] [It]
>   test/e2e/framework/framework.go:633
>   Mar 15 10:35:30.486: failed to see DELETED event
>               
>   Unexpected error:
>       <*errors.errorString | 0xc000204240>: {
>           s: "timed out waiting for the condition",
>       }
>       timed out waiting for the condition
>   occurred
>   test/e2e/common/node/pods.go:1054
> ------------------------------
> {"msg":"FAILED [sig-node] Pods should run through the lifecycle of Pods and PodStatus [Conformance]","total":-1,"completed":1,"skipped":25,"failed":1,"failures":["[sig-node] Pods should run through the lifecycle of Pods and PodStatus [Conformance]"]} 
> ```

@rata 

I did a retest, seems Flaky, passes this time.

endocrimes commented:
/assign

SergeyKanzhelev commented:
/assign @ehashman

ipochi commented:
@SergeyKanzhelev @ehashman 

Friendly ping.

endocrimes reviewed (approved):
/release-note-edit
```release-note
The Kubelet's --lock-file and --exit-on-lock-contention flags are now available via the Kubelet config file field lockFile and exitOnLockContention
```

/lgtm

needs an update in test-infra too - as this will break the lock contention job https://github.com/kubernetes/test-infra/blob/master/config/jobs/kubernetes/sig-node/containerd.yaml#L1315-L1339

Commit:
Update E2E test to use KubeletConfig

In its current incarnation, the e2e test expects lock contention related
features to be enabled via flags.

This change is made so that kubelet updates its configuration via
KubeletConfiguration.

Once the test is complete we revert back to the old
KubeletConfiguration.

Signed-off-by: Imran Pochi <imran@kinvolk.io>

ipochi commented:
@SergeyKanzhelev @endocrimes 

Updated the E2E test so that the test uses KubeletConfiguration to update the kubelet configuration.
Once the test is complete, we revert back to the old kubelet configuration. I think, we can move the test to serial test suite.

Created a PR  https://github.com/kubernetes/test-infra/pull/26185 to remove the flags from the job as well.

ipochi commented:
/retest

ipochi commented:
/test pull-kubernetes-integration
/test pull-kubernetes-verify

endocrimes commented:
/release-note none

ipochi commented:
> @SergeyKanzhelev @endocrimes
> 
> Updated the E2E test so that the test uses KubeletConfiguration to update the kubelet configuration. Once the test is complete, we revert back to the old kubelet configuration. I think, we can move the test to serial test suite.
> 
> Created a PR [kubernetes/test-infra#26185](https://github.com/kubernetes/test-infra/pull/26185) to remove the flags from the job as well.

@SergeyKanzhelev @endocrimes
friendly ping.

ipochi commented:
> > @SergeyKanzhelev @endocrimes
> > Updated the E2E test so that the test uses KubeletConfiguration to update the kubelet configuration. Once the test is complete, we revert back to the old kubelet configuration. I think, we can move the test to serial test suite.
> > Created a PR [kubernetes/test-infra#26185](https://github.com/kubernetes/test-infra/pull/26185) to remove the flags from the job as well.
> 
> @SergeyKanzhelev @endocrimes friendly ping.

Hi @SergeyKanzhelev @endocrimes friendly ping.

SergeyKanzhelev commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -445,6 +445,15 @@ type KubeletConfiguration struct {
+	// LockFilePath is the path that kubelet will use to as a lock file.
+	// It uses this file as a lock to synchronize with other kubelet processes
+	// that may be running.
+	LockFilePath string
```
I suggest to call it `LockFile` without the `Path` to match the `TLSCertFile`. Or `KubeletSynchronizationLockFile`.

SergeyKanzhelev commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -445,6 +445,15 @@ type KubeletConfiguration struct {
+	// LockFilePath is the path that kubelet will use to as a lock file.
+	// It uses this file as a lock to synchronize with other kubelet processes
+	// that may be running.
+	LockFilePath string
```
I suggest to call it `LockFile` without the `Path` to match the `TLSCertFile`. Or `KubeletSynchronizationLockFile`.

SergeyKanzhelev replied:
cc: @kubernetes/api-reviewers for naming

SergeyKanzhelev commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -445,6 +445,15 @@ type KubeletConfiguration struct {
+	// ExitOnLockContention signifies to the kubelet that it is running
+	// in "bootstrap" mode. This requires that 'LockFilePath' has been set.
+	// This will cause the kubelet to listen to inotify events on the lock file,
+	// releasing it and exiting when another process tries to open that file.
```
this description doesn't say what is the behavior when the value is set to `false`. Can you please extend the description

SergeyKanzhelev commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -445,6 +445,15 @@ type KubeletConfiguration struct {
+	// ExitOnLockContention signifies to the kubelet that it is running
+	// in "bootstrap" mode. This requires that 'LockFilePath' has been set.
+	// This will cause the kubelet to listen to inotify events on the lock file,
+	// releasing it and exiting when another process tries to open that file.
```
should this be implied when `LockFilePath` is set? Or there is a scenario when lock file is set and this value is `false`?

cc: @kubernetes/api-reviewers

SergeyKanzhelev replied:
should this be implied when `LockFilePath` is set? Or there is a scenario when lock file is set and this value is `false`?

cc: @kubernetes/api-reviewers

SergeyKanzhelev commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -445,6 +445,15 @@ type KubeletConfiguration struct {
 	// registerNode enables automatic registration with the apiserver.
 	// +optional
 	RegisterNode bool
+	// LockFilePath is the path that kubelet will use to as a lock file.
```
specify is this is a full path or relative path for clarity

SergeyKanzhelev replied:
specify is this is a full path or relative path for clarity

rata commented:
/remove-lifecycle stale

@ipochi friendly ping? :)

dims commented:
If you still need this PR then please rebase, if not, please close the PR
